nav2_costmap_2d/plugins/plugin_container_layer.cpp

```diff
@@ -0,0 +1,235 @@
+
+  dyn_params_handler_ = node->add_on_set_parameters_callback(
+    std::bind(
+      &PluginContainerLayer::dynamicParametersCallback,
+      this,
+      std::placeholders::_1));
+
+
+  plugin_types_.resize(plugin_names_.size());
+
+  for (size_t i = 0; i < plugin_names_.size(); ++i) {
+    plugin_types_[i] = nav2_util::get_plugin_type_param(node, name_ + "." + plugin_names_[i]);
+  }
+
+  // Then load and add the plug-ins to the costmap
+  for (unsigned int i = 0; i < plugin_names_.size(); ++i) {
+    std::shared_ptr<Layer> plugin = plugin_loader_.createSharedInstance(plugin_types_[i]);
+    addPlugin(plugin, plugin_names_[i]);
+  }
+
+  default_value_ = nav2_costmap_2d::FREE_SPACE;
+
+  PluginContainerLayer::matchSize();
+  current_ = true;
+}
+
+void PluginContainerLayer::addPlugin(std::shared_ptr<Layer> plugin, std::string layer_name)
+{
+  plugins_.push_back(plugin);
+  auto node = node_.lock();
+  plugin->initialize(layered_costmap_, name_ + "." + layer_name, tf_, node, callback_group_);
+}
+
+void PluginContainerLayer::updateBounds(
+  double robot_x,
+  double robot_y,
+  double robot_yaw,
+  double * min_x,
+  double * min_y,
+  double * max_x,
+  double * max_y)
+{
+  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin(); plugin != plugins_.end();
+    ++plugin)
+  {
+    (*plugin)->updateBounds(robot_x, robot_y, robot_yaw, min_x, min_y, max_x, max_y);
+  }
+}
+
+void PluginContainerLayer::updateCosts(
+  nav2_costmap_2d::Costmap2D & master_grid,
+  int min_i,
+  int min_j,
+  int max_i,
+  int max_j)
+{
+  std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
+  if (!enabled_) {
+    return;
+  }
+
+  resetMap(min_i, min_j, max_i, max_j);
+
+  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin(); plugin != plugins_.end();
+    ++plugin)
+  {
+    (*plugin)->updateCosts(*this, min_i, min_j, max_i, max_j);
+  }
+
+  switch (combination_method_) {
+    case CombinationMethod::Overwrite:
+      updateWithOverwrite(master_grid, min_i, min_j, max_i, max_j);
+      break;
+    case CombinationMethod::Max:
+      updateWithMax(master_grid, min_i, min_j, max_i, max_j);
+      break;
+    case CombinationMethod::MaxWithoutUnknownOverwrite:
+      updateWithMaxWithoutUnknownOverwrite(master_grid, min_i, min_j, max_i, max_j);
+      break;
+    default:  // Nothing
+      break;
+  }
+
+  current_ = true;
+}
+
+void PluginContainerLayer::activate()
+{
+  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin(); plugin != plugins_.end();
+    ++plugin)
+  {
+    (*plugin)->activate();
+  }
+}
+
+void PluginContainerLayer::deactivate()
+{
+  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin(); plugin != plugins_.end();
+    ++plugin)
+  {
+    (*plugin)->deactivate();
+  }
+}
+
+void PluginContainerLayer::reset()
+{
+  for (vector<std::shared_ptr<Layer>>::iterator plugin = plugins_.begin(); plugin != plugins_.end();
+    ++plugin)
+  {
+    (*plugin)->reset();
+  }
+
```

Comment:
Does the internal layer need to have some update for reset? I imagine so. Any other of these functions?

Reply:
good catch, added resetMap() in here